### ISSUE TITLE
feature requests for RNNs

### ISSUE THREAD

**Issue by CarloLucibello** [FluxML/Flux.jl#2514]
After the redesign in #2500, here is a list of potential improvements for recurrent layers and recurrent cells

- [ ] add an option in constructors to have trainable initial state
- [ ] use cuDNN
- [ ] implement the num_layers argument for stacked RNNs
- [ ] add dropout
- [ ] add bidirectional option

**Comment by MartinuzziFrancesco** [comment]
Could it be possible to add to the list the option to use different initializers for the input matrix and recurrent matrix? This is provided by both [Keras/TF](https://www.tensorflow.org/api_docs/python/tf/keras/layers/LSTMCell) and [Flax](https://flax.readthedocs.io/en/latest/api_reference/flax.nnx/nn/recurrent.html#flax.nnx.nn.recurrent.LSTMCell).

This should be as straightforward as 

```julia
function RNNCell((in, out)::Pair, σ=relu;
    kernel_init = glorot_uniform,
    recurrent_kernel_init = glorot_uniform,
    bias = true)
    Wi = kernel_init(out, in)
    U = recurrent_kernel_init(out, out)
    b = create_bias(Wi, bias, size(Wi, 1))
    return RNNCell(σ, Wi, U, b)
end
```

I can also open a quick PR on this if needed

**Comment by CarloLucibello** [comment]
yes! PR welcome

**Comment by MartinuzziFrancesco** [comment]
Following up on this, should we also have an option to choose the init for the bias?

**Comment by CarloLucibello** [comment]
We don't do it for feedforward layers, if someone wants a non-zero bias can just change it manually in the constructor, `layer.bias .= ...`

**Comment by MartinuzziFrancesco** [comment]
Ehi, a couple of questions on this features request again. Would the `initialstates` function be like the current `function (rnn::RNN)(x::AbstractVecOrMat)` but returning just the state?

would something simple like
```julia
function initialstates(rnn::RNNCell; init_state = zeros)
    state = init_state(size(rnn.Wh, 2))
    return state
end

function initialstates(lstm::LSTMCell; init_state = zeros, init_cstate = zeros)
    state = init_state(size(lstm.Wh, 2))
    cstate = init_cstate(size(lstm.Wh, 2))
    return state, cstate
end
```
suffice or were you looking for something more? Maybe more control on the type would be needed

**Comment by CarloLucibello** [comment]
I would just have
```
initialstates(c::RNNCell) = zeros_like(c.Wh, size(c.Wh, 2)))
initialstates(c:: LSTMCell) = zeros_like(c.Wh, size(c.Wh, 2)), zeros_like(c.Wh, size(c.Wh, 2))
```
If different initializations are needed, we could add an `init_state` to the constructor, but maybe better let the user handle it as a part of the model, for simplicity and flexibility (e.g. making the initial state trainable). I don't have a strong opinion though.

**Comment by MartinuzziFrancesco** [comment]
so this way we would simply do
```julia
function (rnn::RNNCell)(inp::AbstractVecOrMat)
    state = initialstates(rnn)
    return rnn(inp, state)
end
```

to keep compatibility for the current version, right? 

I think your point is good, additionally no other library provides a specific `init` for the state so that's probably a little overkill. I'll push something along these lines later

**Comment by CarloLucibello** [comment]
I think we don't need this additional complexity. A simple stacked rnn can be constructed as a chain:
```julia
stacked_rnn = Chain(LSTM(3 => 3), Dropout(0.5), LSTM(3 => 3))
```
If control of the initial states is also needed, it is not hard to define a custom struct the job:
```julia
struct StackedRNN{L,S}
    layers::L
    states0::S
end

function StackedRNN(d, num_layers)
    layers = [LSTM(d => d) for _ in num_layers]
    states0 = [Flux.initialstates(l) for l in layers]
    return StackedRNN(layers, states0)
end

function (m::StackedRNN)(x)
   for (layer, state0) in zip(rnn.layers, rnn.states0)
       x = layer(x, state0) 
   end
   return x
end
```
I think it is enough to document this in the guide 
https://github.com/FluxML/Flux.jl/blob/master/docs/src/guide/models/recurrence.md

**Comment by MartinuzziFrancesco** [comment]
I assume
>  add an option in the RNN/GRU/LSTM constructor to return the last state, along the output

should be tackled at the `scan` level, maybe something like

```julia
function scan(cell, x, state; return_state = false)
  y = []
  for x_t in eachslice(x, dims = 2)
    yt, state = cell(x_t, state)
    y = vcat(y, [yt])
  end
  if !(return_state)
    return stack(y, dims = 2)
  else
    return stack(y, dims = 2), state
end
```

and then

```julia
struct RNN{M}
  cell::M
  return_state::Bool
end

@layer :noexpand RNN

initialstates(rnn::RNN) = initialstates(rnn.cell)

function RNN((in, out)::Pair, σ = tanh; return_state = false, cell_kwargs...)
  cell = RNNCell(in => out, σ; cell_kwargs...)
  return RNN(cell, return_state)
end

(rnn::RNN)(x::AbstractArray) = rnn(x, initialstates(rnn))

function (m::RNN)(x::AbstractArray, h)
  @assert ndims(x) == 2 || ndims(x) == 3
  # [x] = [in, L] or [in, L, B]
  # [h] = [out] or [out, B]
  return scan(m.cell, x, h; return_state = m.return_state)
end
```

is this the direction you wanted to take this in?

**Comment by CarloLucibello** [comment]
Yes, it could be something like that, although In order to preserve type stability it would be better to embed the `return_state` information as a parameter for the type (it could be a symbol or a bool).

```julia
struct RNN{S, M}
  cell::M
end

function RNN((in, out)::Pair, σ = tanh; return_state = false, cell_kwargs...)
  cell = RNNCell(in => out, σ; cell_kwargs...)
  if return_state
      return RNN{:return_state}(cell)
  else
     return RNN{:no_return_state}(cell)
  end
end

function scan(cell, x, state; return_state = false)
  y = []
  for x_t in eachslice(x, dims = 2)
    yt, state = cell(x_t, state)
    y = vcat(y, [yt])
  end
  return stack(y, dims = 2), state
end

function (m::RNN{:no_return_state)(x::AbstractArray, h)
  return scan(m.cell, x, h; return_state = m.return_state)[1]
end
function (m::RNN{:return_state)(x::AbstractArray, h)
  return scan(m.cell, x, h; return_state = m.return_state)
end
```

**Comment by MartinuzziFrancesco** [comment]
makes sense! this way I also don't think we need the `return_state` in `scan`, so building on your code would something like this work?

```julia

struct RNN{S, M}
  cell::M
end

function RNN((in, out)::Pair, σ = tanh; return_state = false, cell_kwargs...)
  cell = RNNCell(in => out, σ; cell_kwargs...)
     return RNN{return_state, typeof(cell)}(cell)
end

function scan(cell, x, state) # removing  return_state = false)
  y = []
  for x_t in eachslice(x, dims = 2)
    yt, state = cell(x_t, state)
    y = vcat(y, [yt])
  end
  return stack(y, dims = 2), state
end

function (rnn::RNN{false})(inp::AbstractArray, state)
  return first(scan(rnn.cell, inp, state))
end
function (rnn::RNN{true})(inp::AbstractArray, state)
  return scan(rnn.cell, inp, state)
end
```
